CMakeLists.txt

```diff
@@ -1,431 +1,507 @@
-  add_compile_definitions(KAHYPAR_ENABLE_SOED_METRIC)
+  target_compile_definitions(MtKaHyPar-BuildFlags INTERFACE KAHYPAR_ENABLE_SOED_METRIC)
 endif(KAHYPAR_ENABLE_SOED_METRIC)
 
 if(KAHYPAR_ENABLE_STEINER_TREE_METRIC)
-  add_compile_definitions(KAHYPAR_ENABLE_STEINER_TREE_METRIC)
+  target_compile_definitions(MtKaHyPar-BuildFlags INTERFACE KAHYPAR_ENABLE_STEINER_TREE_METRIC)
 endif(KAHYPAR_ENABLE_STEINER_TREE_METRIC)
 
-if(MT_KAHYPAR_DISABLE_BOOST)
-  add_compile_definitions(MT_KAHYPAR_DISABLE_BOOST)
+if(KAHYPAR_DISABLE_HWLOC)
+  target_compile_definitions(MtKaHyPar-BuildFlags INTERFACE KAHYPAR_DISABLE_HWLOC)
 endif()
 
-include_directories(${PROJECT_SOURCE_DIR})
-find_package(Threads REQUIRED)
-message(STATUS "Found Threads: ${CMAKE_THREAD_LIBS_INIT}")
-
-# Include Submodules
-add_subdirectory(external_tools/googletest EXCLUDE_FROM_ALL)
-include_directories(SYSTEM ${gtest_SOURCE_DIR}/include ${gtest_SOURCE_DIR})
-include_directories(SYSTEM ${gtest_SOURCE_DIR}/../googlemock/include ${gtest_SOURCE_DIR}/../googlemock/)
-include_directories(external_tools/kahypar-shared-resources)
-include_directories(external_tools/growt)
-include_directories(external_tools/WHFC)
-include_directories(external_tools/pcg)
-
-if(KAHYPAR_DOWNLOAD_BOOST)
-  # Download Boost
-  execute_process(COMMAND cmake -P ${CMAKE_CURRENT_SOURCE_DIR}/scripts/download_boost.cmake)
-  include_directories(SYSTEM ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/)
-  file(GLOB MINI_BOOST_SOURCES ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/libs/program_options/src/*.cpp)
-
-  add_library(mini_boost STATIC ${MINI_BOOST_SOURCES})
-  set_target_properties(mini_boost PROPERTIES LINKER_LANGUAGE CXX)
-  set(Boost_LIBRARIES mini_boost)
-  set(Boost_INCLUDE_DIRS ${CMAKE_CURRENT_BINARY_DIR}/external_tools/boost/boost/)
-  include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
-else()
-  # Find Boost Program Options Library
-  find_package(Boost 1.69 REQUIRED COMPONENTS program_options)
-  if(Boost_FOUND)
-    include_directories(SYSTEM ${Boost_INCLUDE_DIRS})
-    set(KAHYPAR_INCLUDE_DIRS ${Boost_INCLUDE_DIRS} ${KAHYPAR_INCLUDE_DIRS})
-    set(KAHYPAR_LINK_LIBRARIES ${Boost_LIBRARIES} ${KAHYPAR_LINK_LIBRARIES})
-    message(STATUS "Boost Include: ${Boost_INCLUDE_DIRS}")
-    message(STATUS "Boost Library Dirs: ${Boost_LIBRARY_DIRS}")
-    message(STATUS "Boost Libraries: ${Boost_LIBRARIES}")
-    if(WIN32)
-      add_definitions(-DBOOST_ALL_NO_LIB)
-      add_definitions(-DBOOST_PROGRAM_OPTIONS_DYN_LINK=1)
-    endif()
+
+#################################################################
+## Setup of dependencies                                       ##
+#################################################################
+
+include(FetchContent)
+set(KAHYPAR_TBB_VERSION        v2022.0.0)
+set(KAHYPAR_GOOGLETEST_VERSION v1.15.2)
+set(KAHYPAR_PYBIND11_VERSION   v2.13.6)
+
+message(STATUS "Fetching dependencies...")
+
+# Include header-only dependencies
+FetchContent_Populate(
```

Comment:
Why not keeping them as submodules?

Reply:
FetchContent is preferable to git submodules since it has more features for actual dependency management:
- fetch content supports de-duplicating dependencies which appear multiple times in a dependency tree
- using fetch content allows to more easily replace a dependency with a local version of it, e.g. via find package

Additionally, it allows to keep all dependency handling in one place (i.e., cmake instead of having some dependencies specified by git and some by cmake) and users don't need to remember recursively initializing the submodules when cloning.